### PR TITLE
Feature/implement probability mp.process counts #4943

### PR DESCRIPTION
**Pull request by PietropaoloFrisoni**
### Pull request - issue #4943

**Context: This pull request implements the new feature described in issue #4943 (internal assignment).**

**Description of the Change:**
Implemented the method `process_counts` in the `ProbabilityMP` class to compute a probability measurement from a counts dictionary.

**Benefits:**
Representation of store sampling information in the form of a dictionary counts is much more condensed and memory efficient.

**Possible Drawbacks:** 
The new method should reproduce the results of `process_samples`, implemented in the same class. It has been tested extensively, although I still need to become an expert in quantum computing and circuits. 
The function checks that the provided sampling information is a dictionary matching the format returned by `CountsMP`.

**Related GitHub Issues:** 
#4943 

----------------------------------------------------------------------------------------------------------------------------------------------------------------------

**Further details:** 

First of all, congratulations on the library and your fantastic work! 

I had the time to explore and play with the code, which is quite amazing. I hope to be considered worthy and help you maintain, expand, and improve it.

I am confident you have improvements and interesting suggestions regarding my pull request. I left some questions for you below as a comment on this PR.

Since this is an internal assignment (for which I was given a deadline), I include some information below that might be relevant for your evaluation (time zone: GMT-5):

- *Start time:* December 15th, 9:07 AM 
- *Marked as ready for review on:* December 18th, 3:53 AM
- *End time:* TBD
- *Deadline:* December 29th, 9:07 AM

I am incredibly grateful for your time, and I hope to learn so much from all of you!
I wish you all the best.

### CODE REVIEW

**Comment by PietropaoloFrisoni** [comment]
As I mentioned above, I have a few elementary questions for you:

- I implemented a few checks on the `counts` parameter to verify that it is a dictionary matching the format returned by class:`CountsMP,` defined in `pennylane.measurements.counts.py`. I noticed that there are no similar checks in other methods. Is this a deliberate choice? If that's the case, please tell me if I should remove such checks (or if I should add more of them).
- I understand that the sampling information should be provided so that keys correspond to measurement outcomes in binary strings, like `00010`, etc., and no other digit is allowed. Please let me know if that's not the case, and I'll change the code.
- Finally, just a comment. Since I pushed the last changes, the test associated with `codecov/project` seems stuck, although it was perfectly fine in the previous commit(s). I suspect it is most probably a `codecov` issue.

If you want to test the code and don't want to write something from scratch, I leave you a very basic snippet I used during some of my tests.

I wish you all the best once again!

```
@qml.qnode(qml.device("default.qubit"))
def circuit():
    qml.Hadamard(wires=1)
    qml.Hadamard(wires=2)
    qml.CNOT(wires=[0, 1])
    return qml.sample(wires=(0, 1, 2))


wires = (0, 1)
wire_order = qml.wires.Wires((1, 0))

samples = circuit(shots=100)
counts = {}

for i in range(samples.shape[0]):
    binary_string = "".join(map(str, samples[i, :]))
    counts[binary_string] = counts.get(binary_string, 0) + 1

print(f"samples = {samples}")
print(f"counts = {counts}\n")

prob_old = qml.probs(wires=wires).process_samples(samples, wire_order=wire_order)
prob_new = qml.probs(wires=wires).process_counts(counts, wire_order=wire_order)
(prob_old == prob_new).all()
```

File: pennylane/measurements/probs.py [hunk 1]
```diff
@@ -241,6 +241,43 @@ def process_state(self, state: Sequence[complex], wire_order: Wires):
...
+        assert isinstance(counts, dict), "Sampling information should be a dictionary"
+        assert counts, "Sampling information should not be empty"
+
+        for outcome, occurrence in counts.items():
+            assert isinstance(outcome, str), f"Key {outcome} should be a string"
+            assert all(bit in "01" for bit in outcome), f"Key {outcome} should be binary"
+            assert isinstance(occurrence, int), f"Value {occurrence} should be an integer"
+            assert occurrence > 0, f"Value {occurrence} should be positive"
+
+        assert len(set(len(key) for key in counts.keys())) == 1, "Keys should have the same length"
```
Comment (albi3ro) [inline pennylane/measurements/probs.py:255]: Our validation strategy for PennyLane tends to be:

1. If it's a mistake that a user is likely to make, or someone did make and got confused by
2. If it's a mistake that is particularly hard to trace back and understand what happened

Otherwise, we take the more pythonic route of "let people do what they want and live with the consequences if they're breaking the requirements". This doesn't bog the performance down with a bunch of validation that adds up over time. While assert statements can be optimized away with `-O`, we can't really assume people are doing that most of the time.

But as always, this is an ongoing discussion that morphs over time.

Long story short, let's assume that the counts dictionary looks like it's supposed to.

**Review by albi3ro — COMMENTED** [review]
Sorry about the delay in getting back to you.

This is looking good 🎉  

I left two minor comments, but other than that, this is exactly what I was looking for.

File: pennylane/measurements/probs.py (hunk 1 above)
Comment (mudit2812) [reply pennylane/measurements/probs.py:255]: To add some more context, `MeasurementProcess.process_xyz` are generally developer facing, so it is often okay to leave out such validation in favour of pre-conditions imposed on the function arguments. This is obviously not a good philosophy for user-facing code, but as I said, this is developer-facing (or power-user facing), which allows us to loosen the constraints on the arguments.

Comment (PietropaoloFrisoni) [reply pennylane/measurements/probs.py:255]: Thank you both for clarifying this! 

It makes a lot of sense and explains why there are no similar checks in other methods. I'll remove all checks and asserts as soon as I return home : )

**Comment by PietropaoloFrisoni** [comment]
Hi @albi3ro and @mudit2812.

Thank you so much for your review and your time!


> Sorry about the delay in getting back to you.
> 
No worries, I didn't want to bother you with reminders (especially during Christmas time)



> This is looking good 🎉
> 

> I left two minor comments, but other than that, this is exactly what I was looking for.

Thank you, Christina! 

I pushed a commit with your suggestion and removed the asserts as you both suggested.

If you are not ready to approve the PR, please tell me if there's something else you want me to modify.

Merry Christmas to both of you, as well as to everyone else on the team!

**Comment by mudit2812** [comment]
Regarding the test failure showing up right now. It's unrelated to your PR @PietropaoloFrisoni and there's no need to worry about it. The test validates a finite-shots workflow where the results are non-deterministic. I will resolve this in a different PR. In the mean time, you can re-run CI without having to make new changes by running `git commit --allow-empty -m 'Trigger CI'; git pull; git push`

**Review by mudit2812 — COMMENTED** [review]
These changes look good to me. I'm happy to approve once CI is passing.

**Review by mudit2812 — APPROVED** [review]
Congratulations @PietropaoloFrisoni ! Everything is passing, and the changes look good to me. Happy to get this in.

**Review by albi3ro — APPROVED** [review]
Looks great! 

I hope you've enjoyed the process and the change to learn a little about pennylane 🎉
